### PR TITLE
Disable c10::optional macros

### PR DESCRIPTION
**Pull request by r-barnes**
Test Plan: Sandcastle

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Review by malfet — APPROVED** [review]

File: c10/util/Optional.h [hunk 1]
```diff
@@ -9,6 +9,8 @@
 // up builds.
 
 namespace c10 {
+
+#if !defined(FBCODE_CAFFE2) && !defined(C10_NODEPRECATED)
```
Comment (swolchok) [inline c10/util/Optional.h:13]: is this bc-breaking? if so, needs to be labeled as such. (it doesn't look like C10_NODEPRECATED is defined anywhere, but perhaps it's defined non-greppably)

Comment (r-barnes) [reply c10/util/Optional.h:13]: `C10_NODEPRECATED` is not defined and this is in OSS, so `FBCODE_CAFFE2` is not defined, so the API is still available; thus, it's not bc-breaking.

**Pull request closed without merging**